explorer/ast/ast_rtti.h

```diff
@@ -0,0 +1,143 @@
+  // kinds if we find a reason to do so.
+  template <typename Kind, typename = typename Kind::IsCarbonAstRttiNodeKind>
+  /*implicit*/ AstRttiNodeKind(Kind kind)
+      : AstRttiNodeKind(FromInt(kind.AsInt())) {}
+
+  // Expose the integer value of this node. This is used to set the values of
+  // other enumerations to match, and to implement range checks.
+  using EnumBase::AsInt;
+
+  CARBON_AST_RTTI_KINDS(IGNORE, CARBON_ENUM_CONSTANT_DECLARATION)
+};
+#define CONSTANT_DEFINITION(E) \
+  CARBON_ENUM_CONSTANT_DEFINITION(AstRttiNodeKind, E)
+CARBON_AST_RTTI_KINDS(IGNORE, CONSTANT_DEFINITION)
+#undef CONSTANT_DEFINITION
+
+// Define Kind enumerations for all base classes.
+#define DEFINE_ENUM(C)                                                      \
+  CARBON_DEFINE_RAW_ENUM_CLASS(C##Kind, int) {                              \
+    CARBON_##C##_KINDS(IGNORE, ENUMERATOR)                                  \
+  };                                                                        \
+  template <typename Derived>                                               \
+  class C##KindTemplate : public CARBON_ENUM_BASE_CRTP(C##Kind, Derived) {  \
+   private:                                                                 \
+    using Base = CARBON_ENUM_BASE_CRTP(C##Kind, Derived);                   \
+    friend class AstRttiNodeKind;                                           \
+                                                                            \
+   public:                                                                  \
+    using IsCarbonAstRttiNodeKind = void;                                   \
+                                                                            \
+    C##KindTemplate() = default;                                            \
+                                                                            \
+    /* This type can be explicitly converted from the generic node kind. */ \
+    explicit C##KindTemplate(AstRttiNodeKind base_kind)                     \
+        : Base(Base::FromInt(base_kind.AsInt())) {}                         \
+                                                                            \
+    CARBON_##C##_KINDS(IGNORE, CARBON_INLINE_ENUM_CONSTANT_DEFINITION)      \
+  };                                                                        \
+  class C##Kind : public C##KindTemplate<C##Kind> {                         \
+    using C##KindTemplate<C##Kind>::C##KindTemplate;                        \
+  };
+#define ENUMERATOR(E) E = AstRttiNodeKind::E.AsInt(),
+CARBON_AST_RTTI_KINDS(DEFINE_ENUM, IGNORE)
+#undef DEFINE_ENUM
+#undef ENUMERATOR
+
+// Helper to compute the number of derived classes for a class by repeatedly
+// adding one.
+#define ADD_ONE(C) +1
+#define NUMBER_OF_DERIVED_CLASSES(C) CARBON_##C##_KINDS(IGNORE, ADD_ONE)
+
+// Helper to find the first derived class of a class.
+#define EXPAND(A) A
+#define LPAREN (
+#define RPAREN )
```

Comment:
This is a header file and these defines are of completely unprefixed names. This seems very likely to be broadly incompatible with a wide range of other headers.

Maybe this is tenable in the explorer, but this seems *extremely* risky.

I would be amazed if some of these names aren't defined by `windows.h` for example, making it impossible to include this header afterward.

Reply:
Personally, I don't find it super worthwhile to make our code uglier to *preemptively* work around bad macro hygiene in other people's header files. I believe even windows.h is vastly better at that these days, and generally I think if someone is going to include a bad header and can't get the header fixed, they should undef the leaked macros before including anything else. In a collision, the person including the bad header and our header will typically get a "macro redefined" error rather than silent weirdness, so the cost doesn't seem too high.

But... if you think this is important, I can namespace all of these.

Reply:
OK, so I went and checked. Of these macro names, `<windows.h>` does define `IGNORE`. But it also defines `interface`, which we use as an identifier hundreds of times, including ~100 uses in header files. I think it's not our job to work around problems with `<windows.h>`.

Reply:
Simplified in a way that removes all but one of these macros with short names, and namespaced that one.